tachyon/math/base/semigroups.h

```diff
@@ -216,6 +324,107 @@ class AdditiveSemigroup {
     }
     return ret;
   }
+
+  // scalar: s
+  // bases: [G₀, G₁, ..., Gₙ₋₁]
+  // return: [sG₀, sG₁, ..., sGₙ₋₁]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> ScalingBases(const F& scalar,
+                                            const std::vector<G>& bases) {
+    size_t size = bases.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(bases);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = bases[j].ScalarMul(scalar.ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalars: [s₀, s₁, ..., sₙ₋₁]
+  // base: G
+  // return: [s₀G, s₁G, ..., sₙ₋₁G]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchScalarsWithBase(
+      const std::vector<F>& scalars, const G& base) {
+    size_t size = scalars.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(scalars);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = base.ScalarMul(scalars[j].ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalars: [s₀, s₁, ..., sₙ₋₁]
+  // bases: [G₀, G₁, ..., Gₙ₋₁]
+  // return: [s₀G₀, s₁G₁, ..., sₙ₋₁Gₙ₋₁]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchScalarsToMultiBases(
+      const std::vector<F>& scalars, std::vector<G>& bases) {
+    CHECK_EQ(scalars.size(), bases.size());
+    size_t size = scalars.size();
+    std::vector<ReturnTy> ret(size);
+    size_t num_elems_per_thread = GetNumElementsPerThread(scalars);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < size; i += num_elems_per_thread) {
+      for (size_t j = i; j < i + num_elems_per_thread && j < size; ++j) {
+        ret[j] = bases[j].ScalarMul(scalars[j].ToBigInt());
+      }
+    }
+    return ret;
+  }
+
+  // scalar_powers: [s, s², .., sⁿ⁻¹]
+  // base: G
+  // return: [G, sG, s²G, ..., sⁿ⁻¹G]
+  template <typename F,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  static std::vector<ReturnTy> BatchPowers(size_t size, const F& scalar,
```

Comment:
The comment doesn't match the function signature.
"scalar_powers" -> `something with scalar and size`

How about "CascadeBaseByConst"?

Reply:
How about MulBaseWithScalarPowers? I think "Batch" should not be included in the name for this case since not all cases are parallelized.